salt: fix deployment of `metrics.k8s.io` *APIService*

NicolasT opened the pull request:
This PR eventually fixes the `metrics.k8s.io` `APIService` deployment, but takes a couple of detours to also fix latent bugs discovered by introducing a fix for the primary aim/issue of this patch series...

Fixes: https://github.com/scality/metalk8s/issues/1624

bert-e commented:
# Hello nicolast,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/bugfix/fix-metrics-apiservice` with contents from `w/2.3/bugfix/fix-metrics-apiservice`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/bugfix/fix-metrics-apiservice origin/development/2.4
 $ git merge origin/w/2.3/bugfix/fix-metrics-apiservice
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/bugfix/fix-metrics-apiservice
```

bert-e commented:
# History mismatch



Merge commit #242489469731bb39d620417eaa2fdfc7d89202ef on the integration branch
`w/2.2/bugfix/fix-metrics-apiservice` is merging a branch which is neither the current
branch `bugfix/fix-metrics-apiservice` nor the development branch
`development/2.2`.

It is likely due to a rebase of the branch `bugfix/fix-metrics-apiservice` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/bugfix/fix-metrics-apiservice` with contents from `w/2.3/bugfix/fix-metrics-apiservice`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/bugfix/fix-metrics-apiservice origin/development/2.4
 $ git merge origin/w/2.3/bugfix/fix-metrics-apiservice
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/bugfix/fix-metrics-apiservice
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #63f3211290cee5cda2f265d27578392488375ebf on the integration branch
`w/2.2/bugfix/fix-metrics-apiservice` is merging a branch which is neither the current
branch `bugfix/fix-metrics-apiservice` nor the development branch
`development/2.2`.

It is likely due to a rebase of the branch `bugfix/fix-metrics-apiservice` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT committed:
monitoring: fix `targetPort` name of Prometheus `NodePort`

In a later commit, the handling of `targetPort` will be fixed, after
which the accidental match between the `port` and the `targetPort` (which
is not taken into account right now) would no longer 'work', and instead
the `Service` wouldn't get any `Endpoints`.

See: https://github.com/scality/metalk8s/pull/1630
See: https://github.com/scality/metalk8s/issues/1624

NicolasT committed:
kubernetes: define named ports in control-plane manifests

A later commit will fix the handling of `targetPort` fields in a
`Service` definition, after which the `endpoints` would no longer be
discovered correctly: things only work 'by accident' because the `port`
of these `Service`s match the `targetPort` (numeric) value. However, the
upstream manifests we use contain named port references (rightfully so).

See: https://github.com/scality/metalk8s/pull/1630
See: https://github.com/scality/metalk8s/issues/1624
See: 36020e5f5aecd09e6c3a42c47769270dbcf3205e

NicolasT committed:
salt: properly handle `clusterIP`s in `Service`s

See: https://github.com/scality/metalk8s/pull/1630
See: https://github.com/scality/metalk8s/issues/1624

NicolasT committed:
salt: properly handle `insecureSkipTLSVerify` when deploying an `APIService`

See: https://github.com/scality/metalk8s/pull/1630
See: https://github.com/scality/metalk8s/issues/1624

NicolasT committed:
salt: properly handle `targetPort` and `nodePort` when deploying a `Service`

Also, fix the 'ingress' test-cases which now check for HTTP 404 instead
of 502, which should have raised some alarms from the beginning: a plain
HTTP request hitting `default-backend` results in a 404, not a 502 `Bad
Gateway`: this signaled the backing `Service` could not be properly
reached from the beginning, and given this `Service` uses a `targetPort`
that's different than the `port`, it's susceptible to the same issues as
what's fixed in d784b91d and 36020e5f.

See: https://github.com/scality/metalk8s/pull/1630
See: https://github.com/scality/metalk8s/issues/1624
See: d784b91db3fd215536e785db72c7973a2f793f54
See: 36020e5f5aecd09e6c3a42c47769270dbcf3205e

NicolasT committed:
tests: add test for `metrics.k8s.io` `APIService`

Fixes: https://github.com/scality/metalk8s/issues/1624
See: https://github.com/scality/metalk8s/pull/1630

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/bugfix/fix-metrics-apiservice` with contents from `w/2.3/bugfix/fix-metrics-apiservice`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/bugfix/fix-metrics-apiservice origin/development/2.4
 $ git merge origin/w/2.3/bugfix/fix-metrics-apiservice
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/bugfix/fix-metrics-apiservice
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

NicolasT commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality










*The following options are set:* **approve**

slaperche-scality commented on salt/metalk8s/addons/monitoring/prometheus/deployed.sls:
```diff
@@ -27,7 +27,7 @@ Expose Prometheus:
           protocol: TCP
           node_port: 30222
-          targetPort: api
+          targetPort: web
```
So upstream uses a mix of snake\_case (`node_port`) and camelCase (`targetPort`)? meh…

NicolasT replied:
No. Upstream is definitely 'standard K8s YAML', i.e.`nodePort`. This `deployed.sls` is *not* from upstream, but added by us to create the `NodePort` *Service* for Prometheus. As such:
- `node_port` was used 'to make it work' (I guess)
- Due to the bug in the name handling, `targetPort` was simply ignored
- Things still 'worked' because the (numeric) `targetPort` is 9090, and `targetPort` defaults to `port` if not set (and even though it's set here, it wasn't passed to K8s API, so that one rightfully did what it's supposed to do)
- Due to my fix, `targetPort` is no longer not passed (heh), so it was set to the named port `api`, but our Prometheus *Pod*s don't *have* a port named `api`, so this was incorrect from the beginning.

slaperche-scality commented on salt/_modules/metalk8s_kubernetes.py:
```diff
@@ -1663,6 +1663,10 @@ def __dict_to_service_spec(spec):
                     for port_key, port_value in iteritems(port):
                         if port_key in ("name", "port", "protocol", "node_port", "target_port"):
                             port_kwargs[port_key] = port_value
+                        elif port_key == 'nodePort':
```
So we support several letter casing for the same parameter?

NicolasT replied:
`node_port` is 'standard Salt' (and Python K8s API client). `nodePort` is 'standard K8s YAML'.

NicolasT replied:
So, in a sense this is a bit of an ugly work-around to support both use-cases... The whole interaction with K8s API from Salt ought to be revised anyway (there are some old tickets about this)

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`



Please check the status of the associated issue None.

Goodbye nicolast.
